apps/term-apply/pkg/applicant/manager.go

```diff
@@ -66,91 +60,91 @@ func (a *ApplicantManager) AddApplicant(userID, name, email string, role int) er
 	}
 
 	a.mu.Lock()
-	for i, applicant := range a.applicants {
-		if applicant.userID == userID {
-			if reflect.DeepEqual(newApplicant, applicant) {
+
+	app, err := GetApplication(github, a.dynamodbTable, a.dynamodbIndex)
+	if _, ok := err.(*emptyResultError); ok {
+		log.Printf("Creating new application for applicant %s with (%s, %s, %s)", github, name, email, roleStr)
+		a.mu.Unlock()
+		a.writeChan <- applicationPacket{app: newApplication, updateInPlace: false}
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	if app.github == github {
+		if app.rejected || app.offerGiven {
+			log.Printf(
+				"Found closed application for applicant %s, creating new application (%s, %s, %s)",
+				github,
+				name,
+				email,
+				roleStr,
+			)
+			a.mu.Unlock()
+			a.writeChan <- applicationPacket{app: newApplication, updateInPlace: false}
+		} else {
+			newApplication.appliedDate = app.appliedDate
+			if reflect.DeepEqual(newApplication, app) {
 				log.Printf(
-					"Found existing applicant %s with identical fields, no changes with (%s, %s, %s)",
-					userID,
+					"Found open application for applicant %s with identical fields, no changes with (%s, %s, %s)",
+					github,
 					name,
 					email,
 					roleStr,
 				)
 				a.mu.Unlock()
 			} else {
 				log.Printf(
-					"Found existing applicant %s with updated fields, updating in place with (%s, %s, %s)",
-					userID,
+					"Found open application for applicant %s with updated fields, updating in place with (%s, %s, %s)",
+					github,
 					name,
 					email,
 					roleStr,
 				)
-				a.applicants[i] = newApplicant
 				a.mu.Unlock()
-				a.writeChan <- a.applicants
+				a.writeChan <- applicationPacket{app: newApplication, prevEmail: app.email, updateInPlace: true}
 			}
-			return nil
 		}
+		return nil
+	} else {
+		log.Printf("Creating new application for applicant %s with (%s, %s, %s)", github, name, email, roleStr)
+		a.mu.Unlock()
+		a.writeChan <- applicationPacket{app: newApplication, updateInPlace: false}
+		return nil
 	}
```

Comment:
can we pair on simplifying this? Way too much branching. Will be very hard to maintain. By the time you're into the 3rd else and scrolling to read the entire conditional you might have lost the plot.

Reply:
Definitely :) at the time, I just wanted to get things to work, so I wasn't putting too much thought into how nasty this looks, but I knew that this would be one of the things you point out as soon as you said you had style comments haha

I'll look into a simpler control flow for this. I find switch statements to be a lot cleaner, so maybe I can first figure out the case, then use that in a switch.

Reply:
suggest something more like this as far as control flow:

```
app, err := GetApplication(github, a.dynamodbTable, a.dynamodbIndex)
if _, ok := err.(*emptyResultError); ok {
	log.Printf("Creating new application for applicant %s with (%s, %s, %s)", github, name, email, roleStr)
	a.mu.Unlock()
	a.writeChan <- applicationPacket{app: newApplication, updateInPlace: false}
	return nil
}
if err != nil {
	return err
}

if app.github != github {
	// new candidate
	log.Printf("Creating new application for applicant %s with (%s, %s, %s)", github, name, email, roleStr)
	a.mu.Unlock()
	a.writeChan <- applicationPacket{app: newApplication, updateInPlace: false}
	return nil
}

if app.rejected || app.offerGiven {
	// candidate previously existed and is reapplying
	log.Printf(
		"Found closed application for applicant %s, creating new application (%s, %s, %s)",
		github,
		name,
		email,
		roleStr,
	)
	a.mu.Unlock()
	a.writeChan <- applicationPacket{app: newApplication, updateInPlace: false}
	return nil
}

newApplication.appliedDate = app.appliedDate
if reflect.DeepEqual(newApplication, app) {
	// candidate is reapplying on the same day with the same information
	log.Printf(
		"Found open application for applicant %s with identical fields, no changes with (%s, %s, %s)",
		github,
		name,
		email,
		roleStr,
	)
	a.mu.Unlock()
	return nil
}
// candidate is reapplying on the same day with updated information
log.Printf(
	"Found open application for applicant %s with updated fields, updating in place with (%s, %s, %s)",
	github,
	name,
	email,
	roleStr,
)
a.mu.Unlock()
a.writeChan <- applicationPacket{app: newApplication, prevEmail: app.email, updateInPlace: true}
return nil

```

Reply:
I decided to go with the happy path method where each branch returns instead of a sea of if/else. I also added short comments before each specific case so it should be very clear what's going on now. And if more cases come into play, it's a lot easier to expand this function to include them. (i.e. @sarcasticadmin and I considered adding a lock feature once an applicant starts interviewing, so now it would be as simple as returning after a check for that)

Reply:
awesome!